File: LeetCode/Linked List/Easy/876. Middle of the Linked List/main.cpp
#include <iostream>
using namespace std;

// Definition for singly-linked list.
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    ListNode *middleNode(ListNode *head)
    {
        ListNode *tail = head;
        int length = 0;
        while (tail->next)
        {
            length++;
            tail = tail->next;
        }
        if (length % 2 == 0)
            length = int(length / 2);
        else
            length = int(length / 2 + 1);

        for (int i = 0; i < length; i++)
        {
            head = head->next;
        }
        return head;
    }
};

int main()
{
    Solution s;
    // input  - [1,2,3,4,5,6]
    ListNode *input = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5, new ListNode(6))))));
    ListNode *ans = s.middleNode(input);
    while (ans)
    {
        cout << ans->val << "\t";
        ans = ans->next;
    }
}

File: LeetCode/Linked List/Easy/876. Middle of the Linked List/main.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def middleNode(self, head):
        length = 0
        curr = head
        while(curr):
            length += 1
            curr = curr.next
        # if its odd it will give list starting from mid if its even it wil give list starting from second mid
        # Eg - int(5/2) = 2      and int(6/2) = 3
        for _ in range(0, int(length/2)):
            head = head.next
        return head


obj = Solution()
# input  - [1,2,3,4,5,6]
ans = obj.middleNode(ListNode(1, ListNode(2, ListNode(
    3, ListNode(4, ListNode(5, ListNode(6,)))))))
while(ans):
    print(ans.val)
    ans = ans.next
